File: src/scpn_mirror_core/physics/adiabaticity.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SCPN Mirror Core — fast-ion adiabaticity parameter

"""Fast-ion adiabaticity parameter in the published form.

D. Endrizzi et al., J. Plasma Phys. 89 (2023) 975890501, section 3.6
(after Cohen, Rowlands & Foote 1978b): the first adiabatic invariant fails
when a particle sees a large change of field strength in one gyro-period,
measured by ``alpha = L_B / rho_par`` with ``rho_par = v_par / omega_ci``
and ``L_B = |B| / |grad B|``; the source states that non-adiabatic effects
become significant at ``alpha <= 10``. The gradient scale length is a
declared input at level 0 (a field model belongs to level 2); the parallel
speed is the declared fraction of ``sqrt(2 E_i / m_i)``. A zero fraction
has no parallel gyroradius and is reported as not applicable.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from scpn_mirror_core.errors import DeviceConfigurationError
from scpn_mirror_core.parameters import require_finite, require_positive
from scpn_mirror_core.physics.stability import ion_gyromotion

ADIABATICITY_THRESHOLD = 10.0


def require_fraction(name: str, value: float) -> float:
    """Return ``value`` when it lies in ``[0, 1]``.

    Parameters
    ----------
    name
        Field name reported in the rejection message.
    value
        Value under validation.

    Returns
    -------
    float
        The validated fraction.

    Raises
    ------
    DeviceConfigurationError
        If ``value`` is non-finite or outside ``[0, 1]``.
    """
    require_finite(name, value)
    if not 0.0 <= value <= 1.0:
        raise DeviceConfigurationError(f"{name}: must lie in [0, 1], got {value!r}")
    return value


@dataclass(frozen=True, slots=True)
class Adiabaticity:
    """Adiabaticity parameter of one ion population.

    Parameters
    ----------
    parallel_speed_m_s
        ``f sqrt(2 E_i / m_i)``.
    parallel_gyroradius_m
        ``v_par / omega_ci``; ``None`` when the fraction is zero.
    alpha
        ``L_B / rho_par``; ``None`` when not applicable.
    adiabatic
        ``alpha > 10``; ``None`` when not applicable.
    """

    parallel_speed_m_s: float
    parallel_gyroradius_m: float | None
    alpha: float | None
    adiabatic: bool | None

    def to_record(self) -> dict[str, Any]:
        """Project the parameter to a JSON-serialisable record.

        Returns
        -------
        dict[str, Any]
            Every field under its name.
        """
        return {
            "parallel_speed_m_s": self.parallel_speed_m_s,
            "parallel_gyroradius_m": self.parallel_gyroradius_m,
            "alpha": self.alpha,
            "adiabatic": self.adiabatic,
        }


def adiabaticity(
    field_gradient_scale_length_m: float,
    parallel_velocity_fraction: float,
    ion_energy_kev: float,
    ion_mass_amu: float,
    charge_number: float,
    field_t: float,
) -> Adiabaticity:
    """Evaluate the adiabaticity parameter.

    Parameters
    ----------
    field_gradient_scale_length_m
        ``L_B``; strictly positive.
    parallel_velocity_fraction
        ``v_par / v`` in ``[0, 1]``.
    ion_energy_kev
        Ion energy; strictly positive.
    ion_mass_amu
        Ion mass in proton masses; strictly positive.
    charge_number
        Ion charge ``Z``; strictly positive.
    field_t
        Field where the parameter is evaluated; strictly positive.

    Returns
    -------
    Adiabaticity
        The parameter and its disposition.

    Raises
    ------
    DeviceConfigurationError
        If an input is invalid.
    """
    require_positive("field_gradient_scale_length_m", field_gradient_scale_length_m)
    fraction = require_fraction(
        "parallel_velocity_fraction", parallel_velocity_fraction
    )
    require_positive("ion_energy_kev", ion_energy_kev)
    require_positive("ion_mass_amu", ion_mass_amu)
    require_positive("charge_number", charge_number)
    require_positive("field_t", field_t)
    speed, frequency, _ = ion_gyromotion(
        ion_energy_kev, ion_mass_amu, charge_number, field_t
    )
    parallel_speed = fraction * speed
    if fraction == 0.0:
        return Adiabaticity(
            parallel_speed_m_s=parallel_speed,
            parallel_gyroradius_m=None,
            alpha=None,
            adiabatic=None,
        )
    parallel_gyroradius = parallel_speed / frequency
    alpha = field_gradient_scale_length_m / parallel_gyroradius
    return Adiabaticity(
        parallel_speed_m_s=parallel_speed,
        parallel_gyroradius_m=parallel_gyroradius,
        alpha=alpha,
        adiabatic=alpha > ADIABATICITY_THRESHOLD,
    )

File: rust/src/adiabaticity.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SCPN Mirror Core — fast-ion adiabaticity kernels

//! Fast-ion adiabaticity parameter (Endrizzi et al. 2023, section 3.6),
//! operation-for-operation identical to
//! `scpn_mirror_core.physics.adiabaticity`.

use crate::stability::ion_gyromotion;

/// `alpha` threshold of the disposition.
pub const ADIABATICITY_THRESHOLD: f64 = 10.0;

/// Adiabaticity quantities.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Adiabaticity {
    /// `f sqrt(2 E_i / m_i)`.
    pub parallel_speed_m_s: f64,
    /// `v_par / omega_ci`; `None` at zero fraction.
    pub parallel_gyroradius_m: Option<f64>,
    /// `L_B / rho_par`; `None` at zero fraction.
    pub alpha: Option<f64>,
    /// `alpha > 10`; `None` at zero fraction.
    pub adiabatic: Option<bool>,
}

/// Evaluate the parameter.
#[must_use]
pub fn adiabaticity(
    field_gradient_scale_length_m: f64,
    parallel_velocity_fraction: f64,
    ion_energy_kev: f64,
    ion_mass_amu: f64,
    charge_number: f64,
    field_t: f64,
) -> Adiabaticity {
    let (speed, frequency, _) =
        ion_gyromotion(ion_energy_kev, ion_mass_amu, charge_number, field_t);
    let parallel_speed = parallel_velocity_fraction * speed;
    if parallel_velocity_fraction == 0.0 {
        return Adiabaticity {
            parallel_speed_m_s: parallel_speed,
            parallel_gyroradius_m: None,
            alpha: None,
            adiabatic: None,
        };
    }
    let parallel_gyroradius = parallel_speed / frequency;
    let alpha = field_gradient_scale_length_m / parallel_gyroradius;
    Adiabaticity {
        parallel_speed_m_s: parallel_speed,
        parallel_gyroradius_m: Some(parallel_gyroradius),
        alpha: Some(alpha),
        adiabatic: Some(alpha > ADIABATICITY_THRESHOLD),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn zero_fraction_is_not_applicable_and_alpha_scales_with_field_squared() {
        let none = adiabaticity(1.0, 0.0, 10.0, 2.0, 1.0, 1.0);
        assert!(none.alpha.is_none() && none.adiabatic.is_none());
        let low = adiabaticity(1.0, 0.5, 10.0, 2.0, 1.0, 1.0).alpha.unwrap();
        let high = adiabaticity(1.0, 0.5, 10.0, 2.0, 1.0, 2.0).alpha.unwrap();
        assert!((high / low - 2.0).abs() <= 1.0e-12);
    }
}
